### PR TITLE
Avoid cloning in `log::simplify` and `power::simplify`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/tinfoil-knight/arrow-datafusion/pull/1
Closes https://github.com/apache/arrow-datafusion/issues/10003

Follow on to https://github.com/apache/arrow-datafusion/pull/9983
## Rationale for this change

in https://github.com/apache/arrow-datafusion/pull/9983 @tinfoil-knight  ported over exp/power to be `ScalarUDF`s 🙏  -- while reviewing the PR I noticed that we could reduce some clones and worked up the code in https://github.com/tinfoil-knight/arrow-datafusion/pull/1

## What changes are included in this PR?
1. Avoid cloning in `log::simplify` and `power::simplify`

### CODE REVIEW

**Review by comphead — APPROVED** [review]
💪 thanks @alamb

File: datafusion/functions/src/math/log.rs [hunk 1]
```diff
@@ -146,51 +151,74 @@ impl ScalarUDFImpl for LogFunc {
...
-            &Expr::Literal(ScalarValue::new_ten(&info.get_data_type(number)?)?);
-        if args.len() == 2 {
-            base = &args[0];
-            number = &args[1];
+        // Args are either
+        // log(number)
+        // log(base, number)
+        let num_args = args.len();
+        if args.len() > 2 {
+            return plan_err!("Expected log to have 1 or 2 arguments, got {num_args}");
```
Comment (Omega359) [inline datafusion/functions/src/math/log.rs:162]: should this be an exec error to match similar patterns elsewhere?

File: datafusion/functions/src/math/log.rs [hunk 2]
```diff
@@ -146,51 +151,74 @@ impl ScalarUDFImpl for LogFunc {
...
-        let mut base =
-            &Expr::Literal(ScalarValue::new_ten(&info.get_data_type(number)?)?);
-        if args.len() == 2 {
-            base = &args[0];
-            number = &args[1];
+        // Args are either
+        // log(number)
+        // log(base, number)
+        let num_args = args.len();
+        if args.len() > 2 {
```
Comment (comphead) [inline datafusion/functions/src/math/log.rs:161]:
```suggestion
        if num_args > 2 {
```

File: datafusion/functions/src/math/log.rs (hunk 1 above)
Comment (comphead) [reply datafusion/functions/src/math/log.rs:162]: lets fix it in follow up. tbh we should unify error messaging for all builtin functions

Comment (alamb) [reply datafusion/functions/src/math/log.rs:162]: I think the idea here is that simplify gets called as part of planning (not execution) so planning makes sense to me. Though I am not sure what practical difference it makes in this case 🤔

**Pull request merged**